File: two_pointers/incremovable_subarrays.py

def count_incremovable_subarrays(numbers: list[int]) -> int:  # LeetCode Q.2972.
    total_numbers = len(numbers)
    max_prefix_end_idx = total_numbers - 1
    for idx in range(1, total_numbers):
        if numbers[idx] <= numbers[idx - 1]:
            max_prefix_end_idx = idx - 1
            break

    min_suffix_start_idx = 0
    for idx in range(total_numbers - 1, 0, -1):
        if numbers[idx] <= numbers[idx - 1]:
            min_suffix_start_idx = idx
            break

    incremovable_subarrays = 1  # Can always delete the entire original array.
    suffix_start_idx = min_suffix_start_idx
    for prefix_end_idx in range(max_prefix_end_idx + 1):
        while suffix_start_idx < total_numbers - 1 and numbers[prefix_end_idx] >= numbers[suffix_start_idx]:
            suffix_start_idx += 1

        if numbers[prefix_end_idx] < numbers[suffix_start_idx]:
            incremovable_subarrays += total_numbers + 1 - suffix_start_idx
            continue

        if prefix_end_idx < total_numbers - 1:  # Can't delete an empty subarray.
            incremovable_subarrays += 1

    if min_suffix_start_idx > 0:  # When original array isn't strictly increasing.
        incremovable_subarrays += total_numbers - min_suffix_start_idx

    return incremovable_subarrays

File: two_pointers/incremovable_subarrays/incremovable_subarrays.cpp
#include <vector>
using namespace std;

long long count_incremovable_subarrays(vector<int> &nums)
{ // LeetCode Q.2972.
    int max_rising_prefix_len = 1;
    int idx = 0;
    while (idx < nums.size() - 1 && nums[idx] < nums[idx + 1])
    {
        max_rising_prefix_len += 1;
        idx += 1;
    }

    if (max_rising_prefix_len == nums.size())
    {
        // Base case: entire original array is increasing.
        return (1 + nums.size()) * nums.size() / 2;
    }

    int max_rising_suffix_len = 1;
    idx = nums.size() - 1;
    while (idx > 0 && nums[idx] > nums[idx - 1])
    {
        max_rising_suffix_len += 1;
        idx -= 1;
    }

    // Must use long long to prevent overflow!
    long long incremovable_subarrays = 1; // Incremovable: entire original array.
    // Incremovable: only non-empty suffixes.
    incremovable_subarrays += max_rising_prefix_len;
    // Incremovable: only non-empty prefixes.
    incremovable_subarrays += max_rising_suffix_len;

    int left_idx = 0;
    int right_idx = nums.size() - max_rising_suffix_len;
    while (left_idx < max_rising_prefix_len && right_idx < nums.size())
    { // Incremovable: non-empty subarray in the middle.
        while (nums[left_idx] >= nums[right_idx] && right_idx < nums.size() - 1)
        {
            right_idx += 1; // Find the next compatible non-empty suffix.
        }

        if (nums[left_idx] >= nums[right_idx])
        {
            break; // No compatible non-empty suffixes for remaining prefixes.
        }

        incremovable_subarrays += nums.size() - right_idx;
        left_idx += 1;
    }

    return incremovable_subarrays;
}
